File: code/programs/go/unix-tools/basename_tool.go
// =========================================================================
// basename — Strip Directory and Suffix from Filenames
// =========================================================================
//
// The `basename` utility strips the directory path and optionally a
// suffix from a filename. It's commonly used in shell scripts to extract
// just the filename from a full path.
//
// # Basic usage
//
//   basename /usr/bin/sort      =>  "sort"
//   basename /home/user/file.txt .txt  =>  "file"
//   basename -s .txt file1.txt file2.txt  =>  "file1\nfile2"
//
// # How it works
//
// The algorithm follows POSIX rules:
//   1. If the string is "//", the result is implementation-defined
//      (we return "/").
//   2. Remove trailing slashes.
//   3. If no slashes remain, the string IS the basename.
//   4. Otherwise, remove everything up to and including the last slash.
//   5. If a suffix is specified and the basename ends with it
//      (and is not equal to it), remove the suffix.
//
// # Multiple mode (-a / -s SUFFIX)
//
// In traditional usage, basename takes NAME [SUFFIX] as two positional
// arguments. With -a, it treats all positional arguments as NAMEs and
// processes each one. The -s flag implies -a and specifies the suffix
// to strip from each name.
//
//   basename -a /usr/bin/sort /usr/bin/head
//   =>  "sort\nhead"
//
//   basename -s .go main.go test.go
//   =>  "main\ntest"
//
// # Architecture
//
//   basename.json (spec)       basename_tool.go (this file)
//   ┌──────────────────┐      ┌────────────────────────────────┐
//   │ flags: -a,-s,-z  │      │ for each name:                 │
//   │ variadic NAME    │─────>│   strip directory              │
//   │ help, version    │      │   strip suffix                 │
//   └──────────────────┘      │   output with terminator       │
//                             └────────────────────────────────┘

package main

import (
	"fmt"
	"io"
	"path/filepath"
	"strings"

	clibuilder "github.com/adhithyan15/coding-adventures/code/packages/go/cli-builder"
)

// =========================================================================
// stripBasename — extract the basename from a path
// =========================================================================
//
// This implements the POSIX basename algorithm:
//   1. Remove trailing slashes (unless the string is all slashes).
//   2. Return everything after the last slash.
//
// We use Go's filepath.Base which handles most edge cases correctly,
// including trailing slashes, multiple slashes, and the root path "/".

func stripBasename(name string, suffix string) string {
	// filepath.Base handles the core logic:
	//   "/usr/bin/sort" => "sort"
	//   "/"             => "/"
	//   "."             => "."
	base := filepath.Base(name)

	// Strip the suffix if specified.
	// The suffix is only removed if:
	//   1. The base ends with the suffix
	//   2. The base is NOT equal to the suffix (avoid stripping everything)
	if suffix != "" && base != suffix && strings.HasSuffix(base, suffix) {
		base = base[:len(base)-len(suffix)]
	}

	return base
}

// =========================================================================
// runBasename — the testable core of the basename tool
// =========================================================================

func runBasename(specPath string, argv []string, stdout io.Writer, stderr io.Writer) int {
	// Step 1: Create the parser from the spec.
	parser, err := clibuilder.NewParser(specPath, argv)
	if err != nil {
		fmt.Fprintf(stderr, "basename: %s\n", err)
		return 1
	}

	// Step 2: Parse the arguments.
	result, err := parser.Parse()
	if err != nil {
		fmt.Fprintf(stderr, "%s\n", err)
		return 1
	}

	// Step 3: Handle the result.
	switch r := result.(type) {

	case *clibuilder.HelpResult:
		fmt.Fprintln(stdout, r.Text)
		return 0

	case *clibuilder.VersionResult:
		fmt.Fprintln(stdout, r.Version)
		return 0

	case *clibuilder.ParseResult:
		// Extract flags.
		multiple := getBool(r.Flags, "multiple")
		zero := getBool(r.Flags, "zero")

		// Extract the suffix flag. If -s is given, it implies -a (multiple).
		suffix := ""
		if v, ok := r.Flags["suffix"]; ok {
			if s, ok := v.(string); ok && s != "" {
				suffix = s
				multiple = true
			}
		}

		// Determine the line terminator.
		terminator := "\n"
		if zero {
			terminator = "\x00"
		}

		// Extract positional arguments.
		names := getStringSlice(r.Arguments, "names")
		if len(names) == 0 {
			fmt.Fprintf(stderr, "basename: missing operand\n")
			return 1
		}

		if multiple {
			// Multiple mode (-a or -s): treat all arguments as names.
			for _, name := range names {
				fmt.Fprint(stdout, stripBasename(name, suffix))
				fmt.Fprint(stdout, terminator)
			}
		} else {
			// Traditional mode: first arg is NAME, second (optional) is SUFFIX.
			name := names[0]
			if len(names) > 1 {
				suffix = names[1]
			}
			fmt.Fprint(stdout, stripBasename(name, suffix))
			fmt.Fprint(stdout, terminator)
		}

		return 0

	default:
		fmt.Fprintf(stderr, "basename: unexpected result type: %T\n", result)
		return 1
	}
}

File: code/programs/rust/unix-tools/src/basename_tool.rs
//! # basename — Strip Directory and Suffix from Filenames
//!
//! This module implements the business logic for the `basename` command.
//! The `basename` utility strips the directory portion (and optionally
//! a suffix) from a pathname, leaving just the final component.
//!
//! ## How It Works
//!
//! The algorithm follows POSIX rules:
//!
//! ```text
//!     Input: "/usr/local/bin/program.sh"
//!
//!     Step 1: Remove trailing slashes
//!             "/usr/local/bin/program.sh" → same (no trailing /)
//!
//!     Step 2: Remove everything up to and including the last /
//!             "program.sh"
//!
//!     Step 3: If a suffix is provided, remove it from the end
//!             suffix = ".sh" → "program"
//! ```
//!
//! ## Edge Cases
//!
//! ```text
//!     Input           Result      Explanation
//!     ──────────────  ──────────  ────────────────────────
//!     "/"             "/"         Root is its own basename
//!     "//"            "/"         Multiple slashes → root
//!     "hello"         "hello"     No directory to strip
//!     "hello/"        "hello"     Trailing slash removed
//!     ""              ""          Empty string → empty
//! ```

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/// Strip the directory portion from a path and optionally remove a suffix.
///
/// This implements the POSIX `basename` algorithm:
///
/// 1. If the path is empty, return empty string
/// 2. Remove all trailing slashes (unless the path is entirely slashes)
/// 3. Remove everything before and including the last slash
/// 4. If a suffix is given and the name ends with it (and the name
///    is not equal to the suffix), remove it
///
/// # Examples
///
/// ```text
///     strip_basename("/usr/bin/sort", None)         → "sort"
///     strip_basename("/usr/bin/sort", Some(".rs"))   → "sort"
///     strip_basename("program.sh", Some(".sh"))      → "program"
///     strip_basename("/", None)                      → "/"
/// ```
pub fn strip_basename(path: &str, suffix: Option<&str>) -> String {
    // --- Step 1: Handle empty path ---
    if path.is_empty() {
        return String::new();
    }

    // --- Step 2: Handle all-slashes path ---
    // If the path is nothing but slashes, the basename is "/".
    if path.chars().all(|c| c == '/') {
        return "/".to_string();
    }

    // --- Step 3: Remove trailing slashes ---
    let trimmed = path.trim_end_matches('/');

    // --- Step 4: Find the last slash and take everything after it ---
    let base = match trimmed.rfind('/') {
        Some(pos) => &trimmed[pos + 1..],
        None => trimmed,
    };

    // --- Step 5: Remove suffix if provided ---
    // The suffix is only removed if:
    //   - The base actually ends with the suffix
    //   - The base is not equal to the suffix (POSIX rule)
    let result = match suffix {
        Some(suf) if !suf.is_empty() && base.ends_with(suf) && base != suf => {
            &base[..base.len() - suf.len()]
        }
        _ => base,
    };

    result.to_string()
}

// ---------------------------------------------------------------------------
// Unit tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_path() {
        assert_eq!(strip_basename("/usr/bin/sort", None), "sort");
    }

    #[test]
    fn with_suffix() {
        assert_eq!(strip_basename("program.sh", Some(".sh")), "program");
    }

    #[test]
    fn suffix_not_matching() {
        assert_eq!(strip_basename("program.sh", Some(".txt")), "program.sh");
    }

    #[test]
    fn root_path() {
        assert_eq!(strip_basename("/", None), "/");
    }

    #[test]
    fn multiple_slashes() {
        assert_eq!(strip_basename("///", None), "/");
    }

    #[test]
    fn trailing_slash() {
        assert_eq!(strip_basename("/usr/bin/", None), "bin");
    }

    #[test]
    fn no_directory() {
        assert_eq!(strip_basename("hello", None), "hello");
    }

    #[test]
    fn empty_string() {
        assert_eq!(strip_basename("", None), "");
    }

    #[test]
    fn suffix_equals_basename() {
        // POSIX: if name equals suffix, don't strip
        assert_eq!(strip_basename(".sh", Some(".sh")), ".sh");
    }

    #[test]
    fn deep_path_with_suffix() {
        assert_eq!(
            strip_basename("/home/user/docs/report.pdf", Some(".pdf")),
            "report"
        );
    }
}
